org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/generators/AbstractXML2GrammarGenerator.java

```diff
@@ -0,0 +1,164 @@
+
+	private String doGenerate(Document document, SharedSettings sharedSettings, T generatorSettings) {
+		// Create the generic grammar information from the XML source document.
+		Grammar grammar = createGrammar(document, isFlat());
+		XMLBuilder builder = new XMLBuilder(sharedSettings, "", "");
+		// Generate the grammar content from the grammar information.
+		generate(grammar, generatorSettings, builder);
+		return builder.toString();
+	}
+
+	protected boolean isFlat() {
+		return false;
+	}
+
+	/**
+	 * Generate the grammar content from the given grammr information into the given
+	 * builder.
+	 * 
+	 * @param grammar         the grammar information.
+	 * @param grammarSettings the grammar settings
+	 * @param out             the XML builder to update.
+	 */
+	protected abstract void generate(Grammar grammar, T grammarSettings, XMLBuilder out);
+
+	private Grammar createGrammar(Document document, boolean flat) {
+		Grammar grammar = new Grammar();
+		// Update default namespace
+		String defaultNamespace = null;
+		Element documentElement = document.getDocumentElement();
+		if (documentElement != null) {
+			defaultNamespace = document.getDocumentElement().getAttribute(DOMAttr.XMLNS_ATTR);
+		}
+		grammar.setDefaultNamespace(defaultNamespace);
+		// Update elements information
+		fillElements(document, grammar, grammar, flat);
+		return grammar;
+	}
+
+	private static void fillElements(Node node, Grammar grammar, ContainerDeclaration parent, boolean flat) {
+		if (!node.hasChildNodes()) {
+			return;
+		}
+		NodeList children = node.getChildNodes();
+
+		// Parent
+		if (parent instanceof ElementDeclaration) {
+			List<String> tags = new ArrayList<>(); 
+			ElementDeclaration parentDecl = (ElementDeclaration) parent;
+			for (int i = 0; i < children.getLength(); i++) {
+				Node child = children.item(i);
+				if (child.getNodeType() == Node.ELEMENT_NODE) {
+					Element element = (Element) child;
+					String localName = element.getLocalName();
+					if (!StringUtils.isEmpty(localName)) {
+						tags.add(localName);
+					}
+				}
+			}
+			parentDecl.addChildHierarchy(tags);
+
+			// ((ElementDeclaration) container).addTagName(element.getTagName());
+		}
+
+		for (int i = 0; i < children.getLength(); i++) {
+			Node child = children.item(i);
+			if (child.getNodeType() == Node.ELEMENT_NODE) {
+				Element element = (Element) child;
+				ElementDeclaration elementDecl = getElementDecl(grammar, parent, flat, element);
+				// Update count occurrences
+				elementDecl.incrementOccurrences();
+				// Update has text
+				if (!elementDecl.hasCharacterContent()) {
+					elementDecl.setHasCharacterContent(hasCharacterContent(element));
+				}
+				// Update attributes
+				NamedNodeMap attributes = element.getAttributes();
+				if (attributes != null) {
+					for (int j = 0; j < attributes.getLength(); j++) {
+						Attr attr = (Attr) attributes.item(i);
+						elementDecl.getAttribute(attr.getName());

```

Comment:
Maybe we could have a method `ElementDeclaration#addAttribute` instead? Just an idea

Reply:
I need to improve attributes cardinality, so please keep like this to get the attribute declaration.